blockchain/process_test.go

```diff
@@ -1172,3 +1172,708 @@ func TestProcessLogic(t *testing.T) {
+	params := g.Params()
+	coinbaseMaturity := params.CoinbaseMaturity
+	stakeEnabledHeight := params.StakeEnabledHeight
+	stakeValidationHeight := params.StakeValidationHeight
+
+	// -------------------------------------------------------------------------
+	// Accept all headers in the initial test chain through stake validation
+	// height and the base maturity blocks.
+	//
+	//   genesis -> bfb -> bm0 -> ... -> bm# -> bse0 -> ... -> bse# -> ...
+	//
+	//   ... bsv0 -> ... -> bsv# -> bbm0 -> ... -> bbm#
+	// -------------------------------------------------------------------------
+
+	g.AcceptHeader("bfb")
+	for i := uint16(0); i < coinbaseMaturity; i++ {
+		blockName := fmt.Sprintf("bm%d", i)
+		g.AcceptHeader(blockName)
+	}
+	tipHeight := int64(coinbaseMaturity) + 1
+	for i := int64(0); tipHeight < stakeEnabledHeight; i++ {
+		blockName := fmt.Sprintf("bse%d", i)
+		g.AcceptHeader(blockName)
+		tipHeight++
+	}
+	for i := int64(0); tipHeight < stakeValidationHeight; i++ {
+		blockName := fmt.Sprintf("bsv%d", i)
+		g.AcceptHeader(blockName)
+		tipHeight++
+	}
+	for i := uint16(0); i < coinbaseMaturity; i++ {
+		blockName := fmt.Sprintf("bbm%d", i)
+		g.AcceptHeader(blockName)
+	}
+
+	// -------------------------------------------------------------------------
+	// Initial setup of headers and block data needed for the invalidate and
+	// reconsider tests below.
+	// -------------------------------------------------------------------------
+
+	// Accept block headers for the following tree structure:
+	//
+	// Note that the ! below indicates a block that is invalid due to violating
+	// a consensus rule during the contextual checks prior to connection and the
+	// @ indicates a block that is invalid due to violating a consensus rule
+	// during block connection.
+	//
+	// ... -> b1 -> b2 -> ...
+	//
+	// ... -> b3 -> b4  -> b5   -> b6   -> b7  -> b8  -> b9  -> b10  -> b11
+	//          \-> b4b -> b5b! -> b6b  -> b7b -> b8b       \-> b10a
+	//          \-> b4c                |      \-> b8c -> b9c
+	//          |                      \-> b7d -> b8d -> b9d
+	//          |                      \-> b7e -> b8e
+	//          \-> b4f -> b5f  -> b6f  -> b7f -> b8f -> b9f -> b10f
+	//          \-> b4i -> b5i  -> b6i@
+	//
+	g.AcceptHeader("b1")
+	g.AcceptHeader("b2")
+	g.AcceptHeader("b3")
+	g.AcceptHeader("b4")
+	g.AcceptHeader("b4b")
+	g.AcceptHeader("b4c")
+	g.AcceptHeader("b4f")
+	g.AcceptHeader("b4i")
+	g.AcceptHeader("b5")
+	g.AcceptHeader("b5b") // Invalid block, but header valid.
+	g.AcceptHeader("b5f")
+	g.AcceptHeader("b5i")
+	g.AcceptHeader("b6")
+	g.AcceptHeader("b6b")
+	g.AcceptHeader("b6f")
+	g.AcceptHeader("b6i")
+	g.AcceptHeader("b7")
+	g.AcceptHeader("b7b")
+	g.AcceptHeader("b7d")
+	g.AcceptHeader("b7e")
+	g.AcceptHeader("b7f")
+	g.AcceptHeader("b8")
+	g.AcceptHeader("b8b")
+	g.AcceptHeader("b8c")
+	g.AcceptHeader("b8d")
+	g.AcceptHeader("b8e")
+	g.AcceptHeader("b8f")
+	g.AcceptHeader("b9")
+	g.AcceptHeader("b9c")
```

Comment:
Actually `b9d` is not supposed to be there because it would make that branch have the same work as `b9c` and several of the tests use the different amount of work.  I originally had included it, but decided to remove it to simplify the aforementioned cases.  Looks like I forgot to update the comment accordingly, so I've done that now.